### ISSUE TITLE
Canvas memory corruption with full-size widget.

### ISSUE THREAD

**Issue by p0nce** [AuburnSounds/Dplug#563]
- [x] There is currently a memory corruption error if one full-size widget has a Canvas and the plugin is resized. dplug:canvas seemingly need more space after the line. If you draw in an `OwnedImage` with `trailingSamples` >= 3 it works.

**Comment by p0nce** [comment]
Fixing this needs that _renderedBuffer is a `OwnedImage`, but last time we attempted this it had to be reverted, see bad7c32ac0d4459380d2180960d7cc91d0eb3fd2 why exactly?

**Comment by p0nce** [comment]
I think everything was working OK with Windows and DTK, but not on macOS 10.14 with `OwnedImage`. risky bug

**Comment by p0nce** [comment]
In worst case scenario (ClipIt) `dplug:canvas` needs that each output line have 12 trailing bytes not 3.

**Comment by p0nce** [comment]
Something that follows is that the level 0 of the Diffuse and Material mipmaps should have the same caracteristics (trailing bytes) so that `dplug:canvas` works on them.

**Comment by p0nce** [comment]
It is very troubling that x11window.d and cocoawindow.d do **not** read the returned byte pitch, making it a coupling between the  final buffer of GUIGraphics and the windowing. But it shouldn't be a problem now that renderedbuffer is not the final buffer to be passed to Window.

**Comment by p0nce** [comment]
- https://tronche.com/gui/x/xlib/utilities/XCreateImage.html call can be fixed
- https://developer.apple.com/documentation/coregraphics/1455149-cgimagecreate can be fixed too

**Comment by p0nce** [comment]
To repro original issue in ClipIt:
```d
 override void onDrawRaw(ImageRef!RGBA rawMap, box2i[] dirtyRects)
    {
        super.onDrawRaw(rawMap, dirtyRects);

        float w = position.width;
        float h = position.height;

        foreach(dirtyRect; dirtyRects)
        {
            auto cRaw = rawMap.cropImageRef(dirtyRect);
            canvas.initialize(cRaw);
            canvas.translate(-dirtyRect.min.x, -dirtyRect.min.y);

            // gradients have to be recreated for each dirtyRect
            auto gradient = canvas.createLinearGradient(0, 0, 100*1.414, 100*1.414);
            gradient.addColorStop(0.0f, RGBA(255, 50, 128, 255));
            gradient.addColorStop(1.0f, RGBA(128, 128, 128, 0));

            canvas.fillStyle = gradient;
            canvas.beginPath;
            canvas.moveTo(0, 0);
            canvas.lineTo(200, 0);
            canvas.lineTo(0, 200);
            canvas.fill();
        }
    }
```

**Comment by p0nce** [comment]
Fixed in v12.0.6
